File: src/snes/instruction.rs
use getset::CopyGetters;
use std::cmp::Ordering;
use std::hash::{Hash, Hasher};
use std::rc::Rc;

use crate::analysis::Analysis;
use crate::snes::hardware_registers::HARDWARE_REGISTERS;
use crate::snes::opcodes::{AddressMode, Op, ARGUMENT_SIZES, OPCODES};
use crate::snes::state::StateRegister;

/// Categories of instructions.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum InstructionType {
    Branch,
    Call,
    Interrupt,
    Other,
    Jump,
    Pop,
    Push,
    Return,
    SepRep,
}

/// Structure representing an instruction.
#[derive(Copy, Clone, CopyGetters, Debug, Eq)]
pub struct Instruction {
    /// The address of the instruction.
    #[getset(get_copy = "pub")]
    pc: usize,

    /// The address of the subroutine this instruction belongs to.
    #[getset(get_copy = "pub")]
    subroutine: usize,

    /// Processor state in which the instruction is executed.
    state: StateRegister,

    /// The instruction's opcode byte.
    opcode: u8,

    /// The instruction argument (if any).
    _argument: usize,
}

// Implement some traits for Instruction.
impl Hash for Instruction {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.pc.hash(state);
        self.subroutine.hash(state);
        self.state.p().hash(state);
    }
}
impl PartialEq for Instruction {
    fn eq(&self, other: &Self) -> bool {
        self.pc == other.pc
            && self.subroutine == other.subroutine
            && self.state.p() == other.state.p()
    }
}
impl PartialOrd for Instruction {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
impl Ord for Instruction {
    fn cmp(&self, other: &Self) -> Ordering {
        self.pc.cmp(&other.pc)
    }
}

impl Instruction {
    /// Instantiate an instruction.
    pub fn new(pc: usize, subroutine: usize, p: u8, opcode: u8, argument: usize) -> Self {
        Self {
            pc,
            subroutine,
            state: StateRegister::new(p),
            opcode,
            _argument: argument,
        }
    }

    /// Return the name of the instruction's operation.
    pub fn name(&self) -> String {
        let name: &str = self.operation().into();
        name.to_lowercase()
    }

    /// Return the instruction's operation.
    pub fn operation(&self) -> Op {
        OPCODES[self.opcode as usize].0
    }

    /// Return the instruction's address mode.
    pub fn address_mode(&self) -> AddressMode {
        OPCODES[self.opcode as usize].1
    }

    /// Category of the instruction.
    pub fn typ(&self) -> InstructionType {
        match self.operation() {
            // Call instructions.
            Op::JSR | Op::JSL => InstructionType::Call,

            // Jump instructions.
            Op::JMP | Op::JML | Op::BRA | Op::BRL => InstructionType::Jump,

            // Return instructions.
            Op::RTS | Op::RTL | Op::RTI => InstructionType::Return,

            // Interrupt instructions.
            Op::BRK => InstructionType::Interrupt,

            // SEP/REP instructions.
            Op::SEP | Op::REP => InstructionType::SepRep,

            // Pop instructions.
            Op::PLA | Op::PLB | Op::PLD | Op::PLP | Op::PLX | Op::PLY => InstructionType::Pop,

            // Push instructions.
            Op::PEA
            | Op::PEI
            | Op::PER
            | Op::PHA
            | Op::PHB
            | Op::PHD
            | Op::PHK
            | Op::PHP
            | Op::PHX
            | Op::PHY => InstructionType::Push,

            // Branch instructions.
            Op::BCC | Op::BCS | Op::BEQ | Op::BMI | Op::BNE | Op::BPL | Op::BVC | Op::BVS => {
                InstructionType::Branch
            }

            // Other instructions.
            _ => InstructionType::Other,
        }
    }

    /// Return the instruction's size.
    pub fn size(&self) -> usize {
        self.argument_size() + 1
    }

    /// Return the instruction's argument size.
    pub fn argument_size(&self) -> usize {
        let address_mode = self.address_mode();
        let size = ARGUMENT_SIZES[address_mode];
        if size != -1 {
            return size as usize;
        }

        match address_mode {
            AddressMode::ImmediateM => self.state.a_size(),
            AddressMode::ImmediateX => self.state.x_size(),
            _ => unreachable!(),
        }
    }

    /// Return the instruction's argument, if any.
    pub fn argument(&self) -> Option<usize> {
        match self.argument_size() {
            0 => None,
            1 => Some(self._argument & 0xFF),
            2 => Some(self._argument & 0xFFFF),
            3 => Some(self._argument & 0xFFFFFF),
            _ => unreachable!(),
        }
    }

    /// Return the instruction's argument as an absolute value, if possible.
    pub fn absolute_argument(&self) -> Option<usize> {
        // No argument.
        if self.argument() == None {
            return None;
        }
        let (pc, size, argument) = (
            self.pc as isize,
            self.size() as isize,
            self.argument().unwrap() as isize,
        );

        match self.address_mode() {
            // Fully specified argument.
            AddressMode::ImmediateM
            | AddressMode::ImmediateX
            | AddressMode::Immediate8
            | AddressMode::AbsoluteLong => Some(argument as usize),

            // Partially specified argument.
            AddressMode::Absolute => {
                if self.is_control() {
                    Some(((pc & 0xFF0000) | argument) as usize)
                } else {
                    None
                }
            }

            // Branches.
            AddressMode::Relative => {
                let argument = (argument as i8) as isize;
                Some((pc + size + argument) as usize)
            }
            AddressMode::RelativeLong => {
                let argument = (argument as i16) as isize;
                Some((pc + size + argument) as usize)
            }

            _ => None,
        }
    }

    /// Return whether this is a control instruction.
    pub fn is_control(&self) -> bool {
        let typ = self.typ();
        typ == InstructionType::Branch
            || typ == InstructionType::Call
            || typ == InstructionType::Jump
            || typ == InstructionType::Return
            || typ == InstructionType::Interrupt
    }

    /// Return whether this instruction modifies A.
    pub fn changes_a(&self) -> bool {
        let op = self.operation();
        op == Op::ADC
            || op == Op::AND
            || op == Op::ASL
            || op == Op::DEC
            || op == Op::EOR
            || op == Op::INC
            || op == Op::LDA
            || op == Op::LSR
            || op == Op::ORA
            || op == Op::PLA
            || op == Op::ROL
            || op == Op::ROR
            || op == Op::SBC
            || op == Op::TDC
            || op == Op::TSC
            || op == Op::TXA
            || op == Op::TYA
            || op == Op::XBA
    }

    /// Return whether this instruction modifies the stack pointer.
    pub fn changes_stack(&self) -> bool {
        let op = self.operation();
        op == Op::TCS || op == Op::TXS
    }

    /// Return the instruction's argument as a string.
    pub fn argument_string(&self) -> String {
        // Return the string corresponding to the argument size.
        let arg = || match self.argument_size() {
            1 => format!("${:02X}", self.argument().unwrap()),
            2 => format!("${:04X}", self.argument().unwrap()),
            _ => format!("${:06X}", self.argument().unwrap()),
        };

        match self.address_mode() {
            AddressMode::Implied => String::from(""),
            AddressMode::ImpliedAccumulator => String::from("a"),

            AddressMode::ImmediateM | AddressMode::ImmediateX | AddressMode::Immediate8 => {
                format!("#{}", arg())
            }

            AddressMode::Relative
            | AddressMode::RelativeLong
            | AddressMode::DirectPage
            | AddressMode::Absolute
            | AddressMode::AbsoluteLong
            | AddressMode::StackAbsolute => arg(),

            AddressMode::DirectPageIndexedX
            | AddressMode::AbsoluteIndexedX
            | AddressMode::AbsoluteIndexedLong => format!("{},x", arg()),

            AddressMode::DirectPageIndexedY | AddressMode::AbsoluteIndexedY => {
                format!("{},y", arg())
            }

            AddressMode::DirectPageIndirect
            | AddressMode::AbsoluteIndirect
            | AddressMode::PeiDirectPageIndirect => format!("({})", arg()),

            AddressMode::DirectPageIndirectLong | AddressMode::AbsoluteIndirectLong => {
                format!("[{}]", arg())
            }

            AddressMode::DirectPageIndexedIndirect | AddressMode::AbsoluteIndexedIndirect => {
                format!("({},x)", arg())
            }

            AddressMode::DirectPageIndirectIndexed => format!("({}),y", arg()),

            AddressMode::DirectPageIndirectIndexedLong => format!("[{}],y", arg()),

            AddressMode::StackRelative => format!("{},s", arg()),

            AddressMode::StackRelativeIndirectIndexed => format!("({},s),y", arg()),

            AddressMode::Move => {
                let arg = self.argument().unwrap();
                format!("${:02X},${:02X}", arg >> 8, arg & 0xFF)
            }
        }
    }

    /// Return an argument alias (a label or hardware register), if any.
    pub fn argument_alias(&self, analysis: Rc<Analysis>) -> Option<String> {
        match self.absolute_argument() {
            Some(arg) => {
                if let Some(hw_register) = HARDWARE_REGISTERS.get_by_right(&arg) {
                    Some(hw_register.to_string())
                } else if self.is_control() {
                    analysis.label(arg, Some(self.subroutine))
                } else {
                    None
                }
            }
            None => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_instruction_lda() {
        let instruction = Instruction::new(0x8000, 0x8000, 0b0000_0000, 0xA9, 0x1234);
        assert_eq!(instruction.name(), "lda");
        assert_eq!(instruction.operation(), Op::LDA);
        assert_eq!(instruction.address_mode(), AddressMode::ImmediateM);
        assert_eq!(instruction.typ(), InstructionType::Other);
        assert_eq!(instruction.argument_size(), 2);
        assert_eq!(instruction.size(), 3);
        assert_eq!(instruction.argument().unwrap(), 0x1234);
        assert_eq!(instruction.absolute_argument().unwrap(), 0x1234);
        assert_eq!(instruction.argument_string(), "#$1234");
        assert!(!instruction.is_control());
    }

    #[test]
    fn test_instruction_brl() {
        let instruction = Instruction::new(0x8000, 0x8000, 0b0000_0000, 0x82, 0xFFFD);

        assert_eq!(instruction.name(), "brl");
        assert_eq!(instruction.operation(), Op::BRL);
        assert_eq!(instruction.address_mode(), AddressMode::RelativeLong);
        assert_eq!(instruction.argument_size(), 2);
        assert_eq!(instruction.size(), 3);
        assert_eq!(instruction.argument().unwrap(), 0xFFFD);
        assert_eq!(instruction.absolute_argument().unwrap(), 0x8000);
        assert_eq!(instruction.argument_string(), "$FFFD");
        assert_eq!(instruction.typ(), InstructionType::Jump);
        assert!(instruction.is_control());
    }
}

File: src/instruction.cpp
#include "instruction.hpp"

#include "analysis.hpp"
#include "hardwareregisters.hpp"
#include "utils.hpp"

using namespace std;

// Constructor.
Instruction::Instruction(InstructionPC pc,
                         SubroutinePC subroutinePC,
                         u8 opcode,
                         u24 argument,
                         State state,
                         Analysis* analysis)
    : analysis{analysis},
      pc{pc},
      subroutinePC{subroutinePC},
      opcode{opcode},
      state{state},
      _argument{argument} {}

// Test constructor.
Instruction::Instruction(u8 opcode) : opcode{opcode} {}

// Name of the instruction's operation.
string Instruction::name() const {
  return OPCODE_NAMES[operation()];
}

// Instruction's operation.
Op Instruction::operation() const {
  return OPCODE_TABLE[opcode].first;
}

// Instruction'a address mode.
AddressMode Instruction::addressMode() const {
  return OPCODE_TABLE[opcode].second;
}

// Category of the instruction.
InstructionType Instruction::type() const {
  switch (operation()) {
    // Call instructions.
    case Op::JSR:
    case Op::JSL:
      return InstructionType::Call;

    // Jump instructions.
    case Op::JMP:
    case Op::JML:
    case Op::BRA:
    case Op::BRL:
      return InstructionType::Jump;

    // Return instructions.
    case Op::RTS:
    case Op::RTL:
    case Op::RTI:
      return InstructionType::Return;

    // Interrupt instructions.
    case Op::BRK:
      return InstructionType::Interrupt;

    // SEP/REP instructions.
    case Op::SEP:
    case Op::REP:
      return InstructionType::SepRep;

    // Pop instructions.
    case Op::PLA:
    case Op::PLB:
    case Op::PLD:
    case Op::PLP:
    case Op::PLX:
    case Op::PLY:
      return InstructionType::Pop;

    // Push instructions.
    case Op::PEA:
    case Op::PEI:
    case Op::PER:
    case Op::PHA:
    case Op::PHB:
    case Op::PHD:
    case Op::PHK:
    case Op::PHP:
    case Op::PHX:
    case Op::PHY:
      return InstructionType::Push;

    // Branch instructions.
    case Op::BCC:
    case Op::BCS:
    case Op::BEQ:
    case Op::BMI:
    case Op::BNE:
    case Op::BPL:
    case Op::BVC:
    case Op::BVS:
      return InstructionType::Branch;

    // Other instructions.
    default:
      return InstructionType::Other;
  }
}

// Whether the instruction modifies A.
bool Instruction::changesA() const {
  auto op = operation();
  return op == Op::ADC || op == Op::AND || op == Op::ASL || op == Op::DEC ||
         op == Op::EOR || op == Op::INC || op == Op::LDA || op == Op::LSR ||
         op == Op::ORA || op == Op::PLA || op == Op::ROL || op == Op::ROR ||
         op == Op::SBC || op == Op::TDC || op == Op::TSC || op == Op::TXA ||
         op == Op::TYA || op == Op::XBA;
}

// Whether this is a control instruction.
bool Instruction::isControl() const {
  switch (type()) {
    case InstructionType::Branch:
    case InstructionType::Call:
    case InstructionType::Jump:
    case InstructionType::Return:
    case InstructionType::Interrupt:
      return true;

    default:
      return false;
  }
}
// Whether this is a SEP/REP instruction.
bool Instruction::isSepRep() const {
  return type() == InstructionType::SepRep;
}

// Instruction size.
size_t Instruction::size() const {
  return argumentSize() + 1;
}

// Instruction's argument size.
size_t Instruction::argumentSize() const {
  if (auto size = ARGUMENT_SIZES[addressMode()]) {
    return *size;
  }

  switch (addressMode()) {
    case AddressMode::ImmediateM:
      return state.sizeA();
    case AddressMode::ImmediateX:
      return state.sizeX();
    default:
      __builtin_unreachable();
  }
}

// Instruction's argument, if any.
optional<u24> Instruction::argument() const {
  switch (argumentSize()) {
    case 0:
      return {};
    case 1:
      return _argument & 0xFF;
    case 2:
      return _argument & 0xFFFF;
    case 3:
      return _argument & 0xFFFFFF;
  }
  __builtin_unreachable();
}

// Instruction's argument as an absolute value, if possible.
optional<u24> Instruction::absoluteArgument() const {
  // No argument.
  auto arg = argument();
  if (!arg.has_value()) {
    return {};
  }

  switch (addressMode()) {
    // Fully specified argument.
    case AddressMode::ImmediateM:
    case AddressMode::ImmediateX:
    case AddressMode::Immediate8:
    case AddressMode::AbsoluteLong:
      return arg;

    // Partially specified argument.
    case AddressMode::Absolute:
      return isControl() ? optional((pc & 0xFF0000) | *arg) : nullopt;

    // Branches.
    case AddressMode::Relative:
      return pc + size() + ((i8)*arg);
    case AddressMode::RelativeLong:
      return pc + size() + ((i16)*arg);

    default:
      return {};
  };
};

// Instruction argument as a label, if any.
optional<Label> Instruction::argumentLabel() const {
  if (analysis != nullptr) {
    if (auto arg = absoluteArgument()) {
      // Subroutine / local label.
      if (isControl()) {
        return analysis->getLabel(*arg, subroutinePC);
      }
    }
  }
  return nullopt;
}

// Instruction's argument as a string.
string Instruction::argumentString(bool aliased) const {
  if (aliased) {
    if (auto absArg = absoluteArgument()) {
      auto label = argumentLabel();
      if (label.has_value()) {
        return label->asArgument();
      }

      auto hwRegister = HARDWARE_REGISTERS.find(*absArg);
      if (hwRegister != HARDWARE_REGISTERS.end()) {
        return "!" + hwRegister->second;
      }
    }
  }

  auto arg = argument();
  auto sz = argumentSize();

  switch (addressMode()) {
    default:
    case Implied:
    case ImpliedAccumulator:
      return "";

    case ImmediateM:
    case ImmediateX:
    case Immediate8:
      return format("#$%0" + to_string(sz * 2) + "X", *arg);

    case Relative:
    case RelativeLong:
    case DirectPage:
    case Absolute:
    case AbsoluteLong:
    case StackAbsolute:
      return format("$%0" + to_string(sz * 2) + "X", *arg);

    case DirectPageIndexedX:
    case AbsoluteIndexedX:
    case AbsoluteIndexedLong:
      return format("$%0" + to_string(sz * 2) + "X,x", *arg);

    case DirectPageIndexedY:
    case AbsoluteIndexedY:
      return format("$%0" + to_string(sz * 2) + "X,y", *arg);

    case DirectPageIndirect:
    case AbsoluteIndirect:
    case PeiDirectPageIndirect:
      return format("($%0" + to_string(sz * 2) + "X)", *arg);

    case DirectPageIndirectLong:
    case AbsoluteIndirectLong:
      return format("[$%0" + to_string(sz * 2) + "X]", *arg);

    case DirectPageIndexedIndirect:
    case AbsoluteIndexedIndirect:
      return format("($%0" + to_string(sz * 2) + "X,x)", *arg);

    case DirectPageIndirectIndexed:
      return format("($%0" + to_string(sz * 2) + "X),y", *arg);

    case DirectPageIndirectIndexedLong:
      return format("[$%0" + to_string(sz * 2) + "X],y", *arg);

    case StackRelative:
      return format("$%02X,s", *arg);

    case StackRelativeIndirectIndexed:
      return format("($%02X,s),y", *arg);

    case Move:
      return format("$%02X,$%02X", *arg >> 8, *arg & 0xFF);
  };
}

// Return the state change caused by this instruction, if any.
optional<StateChange> Instruction::stateChange() const {
  return subroutine()->stateChangeForPC(pc);
}

// Get an assertion for the instruction, if any.
optional<Assertion> Instruction::assertion() const {
  return analysis != nullptr ? analysis->getAssertion(pc, subroutinePC)
                             : nullopt;
}

// Get the jumptable associated with the instruction, if any.
const JumpTable* Instruction::jumpTable() const {
  if (analysis != nullptr) {
    auto search = analysis->jumpTables.find(pc);
    if (search != analysis->jumpTables.end()) {
      return &search->second;
    }
  }
  return nullptr;
}

// Get the instruction's coordinates (PC, subroutine's PC).
PCPair Instruction::pcPair() const {
  return {pc, subroutinePC};
}

// Pointer to the subroutine to which the instruction belongs.
Subroutine* Instruction::subroutine() const {
  if (analysis == nullptr) {
    return nullptr;
  }
  return &analysis->subroutines.at(subroutinePC);
}

// Return the instruction's comment.
string Instruction::comment() const {
  if (analysis != nullptr) {
    auto search = analysis->comments.find(pc);
    if (search != analysis->comments.end()) {
      return search->second;
    }
  }
  return "";
}

// Set the instruction's comment.
void Instruction::setComment(string comment) {
  if (analysis != nullptr) {
    if (comment.empty()) {
      analysis->comments.erase(pc);
    } else {
      analysis->comments.insert_or_assign(pc, comment);
    }
  }
}

// Hash table utils.
bool Instruction::operator==(const Instruction& other) const {
  return pc == other.pc && subroutinePC == other.subroutinePC &&
         state == other.state;
}
size_t hash_value(const Instruction& instruction) {
  size_t seed = 0;
  boost::hash_combine(seed, instruction.pc);
  boost::hash_combine(seed, instruction.subroutinePC);
  boost::hash_combine(seed, instruction.state.p);
  return seed;
}
